sdl_android/src/main/java/com/smartdevicelink/encoder/VirtualDisplayEncoder.java

```diff
@@ -0,0 +1,591 @@
+
+        sdlOutStream = videoStream;
+
+        this.presentationClass = presentationClass;
+
+        setupVideoStreamWriter();
+    }
+
+    public void setStreamParams(int frameRate, int bitrate, int interval, int refresh_rate){
+        this.frame_rate = frameRate;
+        this.bitrate = bitrate;
+        this.interval = interval;
+        this.refresh_rate_ms = refresh_rate;
+    }
+
+    private final static Object lock = new Object();
+    /**
+     * Prepares the encoder and virtual display.
+     */
+    @TargetApi(Build.VERSION_CODES.KITKAT)
+    public void start() throws Exception {
+        synchronized (lock) {
+
+            try {
+                inputSurface = prepareVideoEncoder();
+
+                // Create a virtual display that will output to our encoder.
+                virtualDisplay = mDisplayManager.createVirtualDisplay(TAG,
+                        video_width, video_height, DENSITY, inputSurface, DisplayManager.VIRTUAL_DISPLAY_FLAG_PRESENTATION);
+                Log.d(TAG, "Created virtualDisplay.");
+
+                startEncoder();
+                Log.d(TAG, "Starting encoder.");
+
+                displayPresentation();
+                Log.d(TAG, "Displaying presentation.");
+            } catch (Exception ex) {
+                Log.w(TAG, "Unable to create Virtual Display.");
+                throw new RuntimeException(ex);
+            }
+        }
+    }
+
+    @TargetApi(Build.VERSION_CODES.KITKAT)
+    public void shutDown()
+    {
+        try {
+
+            closeVideoSession();
+            releaseVideoStreamWriter();
+
+            if (mVideoEncoder != null) {
+                mVideoEncoder.stop();
+                mVideoEncoder.release();
+                mVideoEncoder = null;
+            }
+
+            if (virtualDisplay != null) {
+                virtualDisplay.release();
+                virtualDisplay = null;
+            }
+
+            if (inputSurface != null) {
+                inputSurface.release();
+                inputSurface = null;
+            }
+        }
+        catch (Exception ex){
+            Log.e(TAG, "shutDown() failed");
+        }
+    }
+
+    private final static Object closeVideoSessionLock = new Object();
+
+    private void closeVideoSession() {
+
+        synchronized (closeVideoSessionLock) {
+            if (sdlOutStream != null) {
+
+                try {
+                    Log.d(TAG, "Closing sdlOutStream.");
+                    sdlOutStream.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+
+                sdlOutStream = null;
+
+                if (streamWriterThread != null) {
+                    Log.d(TAG, "Clearing output stream.");
+                    streamWriterThread.clearOutputStream();
+
+                    Log.d(TAG, "Clearing byte buffer.");
+                    streamWriterThread.clearByteBuffer();
+                }
+            }
+        }
+    }
+
+    @TargetApi(21)
```

Comment:
Should the TargetApi be placed at the top of the class since we're saying if you aren't at that API you can't use this class?